File: 73.矩阵置零.py
#
# @lc app=leetcode.cn id=73 lang=python3
#
# [73] 矩阵置零
#

# @lc code=start
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        flag_col = False
        row, col = len(matrix), len(matrix[0])
        for i in range(row):
            if matrix[i][0] == 0: flag_col = True
            for j in range(1, col):
                if matrix[i][j] == 0:
                    matrix[i][0] = matrix[0][j] = 0
        
        for i in range(row-1, -1, -1):
            for j in range(col-1, 0, -1):
                if matrix[i][0] == 0 or matrix[0][j] == 0:
                    matrix[i][j] = 0
            
            if flag_col == True: matrix[i][0] = 0
# @lc code=end

File: source/73.矩阵置零.cs
/*
 * @lc app=leetcode.cn id=73 lang=csharp
 *
 * [73] 矩阵置零
 */

// @lc code=start
public class Solution
{
  public void SetZeroes(int[][] matrix)
  {
    HashSet<int> rows = new HashSet<int>();
    HashSet<int> cols = new HashSet<int>();

    for (int i = 0; i < matrix.Length; i++) {
      for (int j = 0; j < matrix[0].Length; j++) {
        if (matrix[i][j] == 0) {
          rows.Add(i);
          cols.Add(j);
        }
      }
    }

    foreach (var row in rows) {
      for (int i = 0; i < matrix[0].Length; i++) {
        matrix[row][i] = 0;
      }
    }

    foreach (var col in cols) {
      for (int i = 0; i < matrix.Length; i++) {
        matrix[i][col] = 0;
      }
    }
  }
}
// @lc code=end

